<a href="https://colab.research.google.com/github/lorischl-otter/DS-Unit-1-Sprint-2-Statistics/blob/master/module1/LS_DS_121_Statistics_Probability_Assignment.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

<img align="left" src="https://lever-client-logos.s3.amazonaws.com/864372b1-534c-480e-acd5-9711f850815c-1524247202159.png" width=200>
<br></br>
<br></br>

## *Data Science Unit 1 Sprint 2 Assignment 1*

# Apply the t-test to real data

Your assignment is to determine which issues have "statistically significant" differences between political parties in this [1980s congressional voting data](https://archive.ics.uci.edu/ml/datasets/Congressional+Voting+Records). The data consists of 435 instances (one for each congressperson), a class (democrat or republican), and 16 binary attributes (yes or no for voting for or against certain issues). Be aware - there are missing values!

Your goals:

1. Load and clean the data (or determine the best method to drop observations when running tests)
2. Using hypothesis testing, find an issue that democrats support more than republicans with p < 0.01
3. Using hypothesis testing, find an issue that republicans support more than democrats with p < 0.01
4. Using hypothesis testing, find an issue where the difference between republicans and democrats has p > 0.1 (i.e. there may not be much of a difference)

Note that this data will involve *2 sample* t-tests, because you're comparing averages across two groups (republicans and democrats) rather than a single group against a null hypothesis.

Stretch goals:

1. Refactor your code into functions so it's easy to rerun with arbitrary variables
2. Apply hypothesis testing to your personal project data (for the purposes of this notebook you can type a summary of the hypothesis you formed and tested)

In [0]:
### YOUR CODE STARTS HERE

# try to write a function that would run t-tests for every issue 


**Goal #1**

In [74]:
!wget "https://archive.ics.uci.edu/ml/machine-learning-databases/voting-records/house-votes-84.data"

--2019-11-05 00:04:30--  https://archive.ics.uci.edu/ml/machine-learning-databases/voting-records/house-votes-84.data
Resolving archive.ics.uci.edu (archive.ics.uci.edu)... 128.195.10.252
Connecting to archive.ics.uci.edu (archive.ics.uci.edu)|128.195.10.252|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 18171 (18K) [application/x-httpd-php]
Saving to: ‘house-votes-84.data.1’

house-votes-84.data 100%[===================>]  17.75K  --.-KB/s    in 0.03s   

2019-11-05 00:04:30 (608 KB/s) - ‘house-votes-84.data.1’ saved [18171/18171]



In [75]:
import pandas as pd
import numpy as np

column_headers = ['party','handicapped-infants','water-project',
                          'budget','physician-fee-freeze', 'el-salvador-aid',
                          'religious-groups','anti-satellite-ban',
                          'aid-to-contras','mx-missile','immigration',
                          'synfuels', 'education', 'right-to-sue','crime','duty-free',
                          'south-africa']

df = pd.read_csv('house-votes-84.data',
                 header=None,
                 names=column_headers,
                 na_values="?")
df.head()

party handicapped-infants water-project  ... crime duty-free south-africa
0  republican                   n             y  ...     y         n            y
1  republican                   n             y  ...     y         n          NaN
2    democrat                 NaN             y  ...     y         n            n
3    democrat                   n             y  ...     n         n            y
4    democrat                   y             y  ...     y         y            y

[5 rows x 17 columns]

In [76]:
df = df.replace({'y': 1, 'n': 0})
df.head()

party  handicapped-infants  ...  duty-free  south-africa
0  republican                  0.0  ...        0.0           1.0
1  republican                  0.0  ...        0.0           NaN
2    democrat                  NaN  ...        0.0           0.0
3    democrat                  0.0  ...        0.0           1.0
4    democrat                  1.0  ...        1.0           1.0

[5 rows x 17 columns]

In [77]:
rep = df[df['party'] == 'republican']
rep.head()

party  handicapped-infants  ...  duty-free  south-africa
0   republican                  0.0  ...        0.0           1.0
1   republican                  0.0  ...        0.0           NaN
7   republican                  0.0  ...        NaN           1.0
8   republican                  0.0  ...        0.0           1.0
10  republican                  0.0  ...        0.0           0.0

[5 rows x 17 columns]

In [78]:
dem = df[df['party'] == 'democrat']
dem.head()

party  handicapped-infants  water-project  ...  crime  duty-free  south-africa
2  democrat                  NaN            1.0  ...    1.0        0.0           0.0
3  democrat                  0.0            1.0  ...    0.0        0.0           1.0
4  democrat                  1.0            1.0  ...    1.0        1.0           1.0
5  democrat                  0.0            1.0  ...    1.0        1.0           1.0
6  democrat                  0.0            1.0  ...    1.0        1.0           1.0

[5 rows x 17 columns]

**Goal #2**

Using hypothesis testing, find an issue that democrats support more than republicans with p < 0.01


In [79]:
dem['budget'].mean()

0.8884615384615384

In [80]:
rep['budget'].mean()

0.13414634146341464

1) Null Hypothesis: There is no difference between rates of support for the "budget" bill between republicans and democrats. <br>
$\bar{x}_{1}$ == $\bar{x}_{2}$, where $\bar{x}_{1}$ is the average republican support, and $\bar{x}_{2}$ is the average democrat support.

2) Alternative Hypothesis: There is a difference between rates of support for the "budget" bill between republicans and democrats. <br>
$\bar{x}_{1}$ $\neq$ $\bar{x}_{2}$


3) Confidence Level: 95%

In [0]:
from scipy.stats import ttest_ind

In [82]:
ttest_ind(rep['budget'], dem['budget'], nan_policy='omit')

Ttest_indResult(statistic=-23.21277691701378, pvalue=2.0703402795404463e-77)

4) T-Statistic: -23.21

5) P-Value = <.01 (2.07e-77)

6) Due to a p-value that is very close to zero, I reject the null hypothesis that republican and democrat support for the 'budget' bill was the same, and conclude that it is something different. $\bar{x}_{1}$ $\neq$ $\bar{x}_{2}$


**Goal #3**

Using hypothesis testing, find an issue that republicans support more than democrats with p < 0.01


In [83]:
rep['physician-fee-freeze'].mean()

0.9878787878787879

In [84]:
dem['physician-fee-freeze'].mean()

0.05405405405405406

1) Null Hypothesis: There is no difference between rates of support for the "physician-fee-freeze" bill between republicans and democrats. <br>
$\bar{x}_{1}$ == $\bar{x}_{2}$, where $\bar{x}_{1}$ is the average republican support, and $\bar{x}_{2}$ is the average democrat support.

2) Alternative Hypothesis: There is a difference between rates of support for the "physician-fee-freze" bill between republicans and democrats. <br>
$\bar{x}_{1}$ $\neq$ $\bar{x}_{2}$


3) Confidence Level: 95%

In [85]:
ttest_ind(rep['physician-fee-freeze'], dem['physician-fee-freeze'], 
          nan_policy="omit")

Ttest_indResult(statistic=49.36708157301406, pvalue=1.994262314074344e-177)

4) T-Statistic: 49.37

5) P-Value = <.01 (1.99e-177)

6) Due to a p-value that is very close to zero, I reject the null hypothesis that republican and democrat support for the 'physician fee freeze' bill was the same, and conclude that it is something different. $\bar{x}_{1}$ $\neq$ $\bar{x}_{2}$

**Goal #4**

Using hypothesis testing, find an issue where the difference between republicans and democrats has p > 0.1 (i.e. there may not be much of a difference)


In [86]:
rep['el-salvador-aid'].mean()

0.9515151515151515

In [87]:
dem['el-salvador-aid'].mean()

0.21568627450980393

In [88]:
rep['aid-to-contras'].mean()

0.15286624203821655

In [89]:
dem['aid-to-contras'].mean()

0.8288973384030418

In [90]:
rep['anti-satellite-ban'].mean()

0.24074074074074073

In [91]:
dem['anti-satellite-ban'].mean()

0.7722007722007722

In [92]:
rep['mx-missile'].mean()

0.11515151515151516

In [93]:
dem['mx-missile'].mean()

0.7580645161290323

In [94]:
rep['right-to-sue'].mean()

0.8607594936708861

In [95]:
dem['right-to-sue'].mean()

0.2896825396825397

In [96]:
rep['crime'].mean()


0.9813664596273292

In [97]:
dem['crime'].mean()

0.35019455252918286

In [98]:
rep['education'].mean()

0.8709677419354839

In [99]:
dem['education'].mean()

0.14457831325301204

In [100]:
rep['religious-groups'].mean()

0.8975903614457831

In [101]:
dem['religious-groups'].mean()

0.47674418604651164

In [102]:
rep['water-project'].mean()

0.5067567567567568

In [103]:
dem['water-project'].mean()

0.502092050209205

1) Null Hypothesis: There is no difference between rates of support for the "water-project" bill between republicans and democrats. <br>
$\bar{x}_{1}$ == $\bar{x}_{2}$, where $\bar{x}_{1}$ is the average republican support, and $\bar{x}_{2}$ is the average democrat support.

2) Alternative Hypothesis: There is a difference between rates of support for the "water-project" bill between republicans and democrats. <br>
$\bar{x}_{1}$ $\neq$ $\bar{x}_{2}$


3) Confidence Level: 95%

In [104]:
ttest_ind(rep['water-project'], dem['water-project'], nan_policy='omit')

Ttest_indResult(statistic=0.08896538137868286, pvalue=0.9291556823993485)

4) T-Statistic: .089

5) P-Value = .929

6) Due to a P-Value of .929, I fail to reject the null hypothesis that republican and democrat support for the 'water project' bill is the same.

**Stretch Goal #1**

Refactor your code into functions so it's easy to rerun with arbitrary variables


In [105]:
rep.head()
rep2 = rep.drop('party', axis=1)
rep2.head()

handicapped-infants  water-project  budget  ...  crime  duty-free  south-africa
0                   0.0            1.0     0.0  ...    1.0        0.0           1.0
1                   0.0            1.0     0.0  ...    1.0        0.0           NaN
7                   0.0            1.0     0.0  ...    1.0        NaN           1.0
8                   0.0            1.0     0.0  ...    1.0        0.0           1.0
10                  0.0            1.0     0.0  ...    1.0        0.0           0.0

[5 rows x 16 columns]

In [106]:
from scipy.stats import ttest_1samp

def t_tests(df):
  for col in list(df):
    return ttest_1samp(df[col], .5, nan_policy='omit')

rep2.apply(t_tests)

/usr/local/lib/python3.6/dist-packages/numpy/core/fromnumeric.py:3506: RuntimeWarning: Degrees of freedom <= 0 for slice
  **kwargs)
/usr/local/lib/python3.6/dist-packages/numpy/core/_methods.py:209: RuntimeWarning: invalid value encountered in double_scalars
  ret = ret.dtype.type(ret / rcount)


TypeError: ignored

In [0]:
def t_tests(df1, df2):
  for 

In [0]:
list(rep)

**Stretch Goal #2**

Apply hypothesis testing to your personal project data (for the purposes of this notebook you can type a summary of the hypothesis you formed and tested)

In [0]:
!wget 'https://archive.ics.uci.edu/ml/machine-learning-databases/census-income-mld/census-income.data.gz'

In [0]:
# column info:
| age						AAGE
| class of worker				ACLSWKR
| industry code					ADTIND
| occupation code				ADTOCC
| adjusted gross income				AGI
| education					AHGA
| wage per hour					AHRSPAY
| enrolled in edu inst last wk			AHSCOL
| marital status				AMARITL
| major industry code				AMJIND
| major occupation code				AMJOCC
| mace						ARACE
| hispanic Origin				AREORGN
| sex						ASEX
| member of a labor union			AUNMEM
| reason for unemployment			AUNTYPE
| full or part time employment stat		AWKSTAT
| capital gains					CAPGAIN
| capital losses				CAPLOSS
| divdends from stocks				DIVVAL
| federal income tax liability			FEDTAX
| tax filer status				FILESTAT
| region of previous residence			GRINREG
| state of previous residence			GRINST
| detailed household and family stat		HHDFMX
| detailed household summary in household	HHDREL
| instance weight				MARSUPWT
| migration code-change in msa			MIGMTR1
| migration code-change in reg			MIGMTR3
| migration code-move within reg		MIGMTR4
| live in this house 1 year ago			MIGSAME
| migration prev res in sunbelt			MIGSUN
| num persons worked for employer		NOEMP
| family members under 18			PARENT
| total person earnings				PEARNVAL
| country of birth father			PEFNTVTY
| country of birth mother			PEMNTVTY
| country of birth self				PENATVTY
| citizenship					PRCITSHP
| total person income				PTOTVAL
| own business or self employed			SEOTR
| taxable income amount				TAXINC
| fill inc questionnaire for veteran's admin	VETQVA
| veterans benefits				VETYN
| weeks worked in year				WKSWORK

In [0]:
column_headers = ['AAGE', 'ACLSWKR', 'ADTIND', 'ADTOCC', 'AGI', 'AHGA',
                  'AHRSPAY', 'AHSCOL', 'AMARITL', 'AMJIND', 'AMJOCC', 'ARACE',
                  'AREORGN', 'ASEX', 'AUNMEM', 'AUNTYPE', 'AWKSTAT', 'CAPGAIN',
                  'CAPLOSS', 'DIVVAL', 'FEDTAX', 'FILESTAT', 'GRINREG',
                  'GRINST', 'HHDREL', 'MARSUPWT', 'MIGMTR1', 'MIGMTR3',
                  'MIGMTR4', 'MIGSAME', 'MIGSUN', 'NOEMP', 'PARENT', 'PEARNVAL',
                  'PEFNTVTY', 'PEMNTVTY', 'PENATVTY', 'PRCITSHP', 'PTOTVAL',
                  'SEOTR', 'TAXINC', 'VETQVA', 'VETYN', 'WKSWORK']

census_income = pd.read_csv('census-income.data.gz', 
                            header=None,
                            names=column_headers,
                            na_values=('?', 'Not in universe'))
census_income.head()

In [0]:
census_income.sample(10)